### PR TITLE
Task3 # PostgreSQL and layered architecture

### PR DESCRIPTION
**Pull request by akuchynski**
Environment variables must be located in .env file (please, see .env.demo).
Sequelize Migration was configured to filled table with predefined users (please, see readme).
All CRUD operations are done using database (PostgreSQL).

### CODE REVIEW

File: api/app.js [hunk 1]
```diff
@@ -1,17 +1,24 @@
 const express = require('express');
 const app = express();
-const User = require('./models/user.js');
-const userRouter = require('./routes/userRouter.js');
+const userRouter = require('./routes/userRouter');
+const dotenv = require('dotenv');
+
+dotenv.config();
```
Comment (artem123sh) [inline api/app.js:6]: Could dotenv.config(); be executed config.js only and removed from other places?

File: api/routes/userRouter.js [hunk 2]
```diff
@@ -1,19 +1,16 @@
...
+const userSchema = require('../schemas/userSchema');
 const validator = require('express-joi-validation').createValidator({});
 
 router.route('/')
     .post(validator.body(userSchema.schema), userController.create)
-    .get(userController.findAll);
-
-router.route('/suggest')
-    .get(userController.suggestUsers);
+    .get(userController.getAll);
```
Comment (artem123sh) [inline api/routes/userRouter.js:9]: Just a small note, we can also use joi validations for query and path params.

File: api/controllers/userController.js [hunk 3]
```diff
@@ -1,71 +1,58 @@
...
-        User.removeUser(user);
-        res.redirect('/users');
-    }
+async function remove(req, res) {
+    await models.user.destroy({
+        where: {
+            id: req.params.id
+        }
+    });
+    res.status(200).end();
```
Comment (artem123sh) [inline api/controllers/userController.js:49]: Superminor: we can use res.sendStatus(200) without end :)

File: api/controllers/userController.js [hunk 4]
```diff
@@ -1,71 +1,58 @@
...
-            req.params.id,
-            req.body.login,
-            req.body.password,
-            req.body.age,
-            req.body.isDeleted
-        );
-        User.updateUser(userData);
-        res.redirect('/users');
+    try {
+        await models.user.update(req.body, {
```
Comment (artem123sh) [inline api/controllers/userController.js:32]: Would it be better to introduce a service which will deal with DB models and have controller independent of DB?

File: api/controllers/userController.js [hunk 5]
```diff
@@ -1,71 +1,58 @@
...
 
-exports.create = (req, res) => {
-    if (!req.body) {
-        return res.status(400).send({
-            message: 'User content can not be empty!'
-        });
+async function create(req, res) {
+    try {
+        await models.user.create(req.body);
+        res.status(201).end();
```
Comment (artem123sh) [inline api/controllers/userController.js:6]: Would be great to return user's Id here.

File: api/controllers/userController.js [hunk 6]
```diff
@@ -1,71 +1,58 @@
...
-exports.suggestUsers = (req, res) => {
-    const users = User.getAutoSuggestUsers(req.query.login, req.query.limit);
-    if (users.length == 0) {
-        res.json({ message: `User list is empty!` });
-    } else {
-        res.json(users);
+async function update(req, res) {
+    const user = await models.user.findByPk(req.params.id);
+    if (!user) {
+        res.status(400).send(`User with id ${req.params.id} not found!`);
```
Comment (artem123sh) [inline api/controllers/userController.js:29]: Looks like status should be 404 :)

File: api/controllers/userController.js [hunk 7]
```diff
@@ -1,71 +1,58 @@
...
-    if (!req.body) {
-        return res.status(400).send({
-            message: 'User content can not be empty!'
-        });
+async function create(req, res) {
+    try {
+        await models.user.create(req.body);
+        res.status(201).end();
+    } catch (error) {
+        console.log(error);
```
Comment (artem123sh) [inline api/controllers/userController.js:8]: Would be better to use the following signature of controller:
async function create(req, res. next) {
try {
//do smth
} catch (e) {
//handle error or return error to pass error to express
next(e) 
}

File: api/app.js (hunk 1 above)
Comment (akuchynski) [reply api/app.js]: the config was changed to preload dotenv

File: api/db/models/user.model.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+const { DataTypes } = require('sequelize');
+module.exports = (sequelize) => {
+    sequelize.define('user', {
+        id: {
+            type: DataTypes.INTEGER,
+            primaryKey: true,
+            autoIncrement: true,
+            allowNull: false
+        },
+        login: {
```
[the remark this answers is not in the document]
Comment (akuchynski) [reply api/db/models/user.model.js:10]: login validation was added

File: api/controllers/userController.js (hunk 6 above)
Comment (akuchynski) [reply api/controllers/userController.js]: the status was changed

File: api/controllers/userController.js (hunk 3 above)
Comment (akuchynski) [reply api/controllers/userController.js]: recommendation was applied

File: api/controllers/userController.js (hunk 5 above)
Comment (akuchynski) [reply api/controllers/userController.js]: the code was modified to fetch user's id after creating

File: api/routes/userRouter.js (hunk 2 above)
Comment (akuchynski) [reply api/routes/userRouter.js:9]: thank for recommendation, joi will be used for query and path params in next tasks.

File: api/controllers/userController.js (hunk 7 above)
Comment (akuchynski) [reply api/controllers/userController.js]: recommendation was applied, 'next' was added to controller functions

File: api/controllers/userController.js (hunk 4 above)
Comment (akuchynski) [reply api/controllers/userController.js]: userService class was added to deal with DB logic

**Pull request merged**